File: src/frontend/src/lib/clientFingerprint.ts
/**
 * A cheap, best-effort device fingerprint sent as `X-Client-Fingerprint` on
 * the plot-suggestion request — the third identity signal
 * `PlotReservationService` combines with `OwnerId` (the primary pin) and IP
 * (a NAT-tolerant cap) to make incognito/private-window `OwnerId` churn from
 * the same device harder: a private window wipes localStorage but not
 * screen size, timezone, or language. Not a security control, just a
 * deterrent at the same trust level as the IP cap — the backend hashes this
 * together with User-Agent/Accept-Language itself.
 *
 * Computed once and memoised; wrapped so a missing `screen`/`Intl` (an
 * unusual embedding context, not real browsers) degrades to an empty string
 * rather than throwing and breaking the request that needs it.
 */
let cached: string | null = null;

export function clientFingerprint(): string {
  if (cached !== null) return cached;

  try {
    const parts = [
      screen.width,
      screen.height,
      window.devicePixelRatio,
      Intl.DateTimeFormat().resolvedOptions().timeZone,
      navigator.language,
      navigator.hardwareConcurrency,
      navigator.platform,
    ];
    cached = parts.join('|');
  } catch {
    cached = '';
  }

  return cached;
}

File: src/backend/src/Bjarnoy.Api/Auth/ClientFingerprint.cs
using System.Security.Cryptography;
using System.Text;

namespace Bjarnoy.Api.Auth;

/// <summary>
/// A coarse, server-side hash approximating "this browser on this device" —
/// the third identity signal <c>PlotReservationService</c>'s abuse caps use
/// alongside <c>OwnerId</c> (the primary pin) and IP (a NAT-tolerant cap).
/// Aimed specifically at incognito/private-window <c>OwnerId</c> churn:
/// a private window wipes localStorage (and so the client's persisted
/// <c>OwnerId</c>) but not its User-Agent, Accept-Language, or the client's
/// own cheap device fingerprint header, so it still hashes to the same value.
/// Not a security control — a determined visitor can still vary these — only
/// a cheap deterrent, same trust level as the IP cap it sits alongside.
/// </summary>
internal static class ClientFingerprint
{
    public const string ClientFingerprintHeaderName = "X-Client-Fingerprint";

    public static string Derive(IHeaderDictionary headers)
    {
        var userAgent = headers["User-Agent"].ToString();
        var acceptLanguage = headers["Accept-Language"].ToString();
        var clientHint = headers[ClientFingerprintHeaderName].ToString();

        var raw = $"{userAgent}\n{acceptLanguage}\n{clientHint}";
        var hash = SHA256.HashData(Encoding.UTF8.GetBytes(raw));
        return Convert.ToHexStringLower(hash);
    }
}
